core/fed_avg_optimizer.py

```diff
@@ -105,12 +129,30 @@ def _handle_job_done(self, dmlresult_obj):
 	def _done_initializing(self, dmlresult_obj):
 		"""
 		"LEVEL 2" Callback for a initialization job that just completed. Returns
-		a DML Job of type train and modifies the current state of the job.
+		a DML Job of type train unless transforming and splitting is not yet done
+		and modifies the current state of the job.
 		"""
 		new_weights = dmlresult_obj.results.get('weights')
 		self._update_weights(new_weights)
-		self.job.job_type = JobTypes.JOB_TRAIN.name
-		return ActionableEventTypes.SCHEDULE_JOB.name, self.job
+		if self.job.job_type == JobTypes.JOB_TRAIN.name:
```

Comment:
So if I understand correctly, you're scheduling an INIT and TRANSFORM_SPLIT job at the same time. When the first job finishes, you set `self.job.job_type = JobTypes.JOB_TRAIN.name`. That way when the other job finishes, it knows the other one already finished and you can start training.

If so, this is a pretty good solution. However I think you should just use separate variables that can be booleans (set to False at start). For example, `_done_initializing` can check a variable called `self.has_finished_transform_split`. If the variable is True, then start training, otherwise set `self.has_finished_initializing` to True.  Vice versa for other method. This might seem nitpicky, but I think it makes your algorithm a lot more readable and easier to understand.

Reply:
Separate variables encounter synchronization issues. Recall that _done_initializing and _done_training are by definition running in different processes. If _done_initializing hits the 'else' and then is preempted by _done_training, which also hits the 'else' because _done_initializing was preempted before it set the variable to True, then we have an issue.
When two processes in Python handle the same variable there's an implicit thread_lock being acquired and released.
See https://en.wikipedia.org/wiki/Lock_(computer_science) for more details on locks.
I can use another variable name, or add to the docstring, to make it easier to understand.

Reply:
I see but could you explain why the same issue doesn't apply with the job type? Specifically, what if both processes hit the else clause before the job type is updated?

Reply:
>When two processes in Python handle the same variable there's an implicit thread_lock being acquired and released.

Reply:
But I'm asking about the case where the first process will check the job type -> move to else, then the second process will check the job type -> move to else, first process sets job type, second process sets job type.

Reply:
Doesn't the lock get released when one process is done with whatever clause/assignment uses that variable?

Reply:
That case doesn't occur because both functions acquire any locks for variables that they are setting at the beginning of the function. So the second process is blocked if both of them are setting the same variable, since it needs that lock. However if they aren't using the same variable, Python sees no need to enforce the lock - after all, if you acquired a lock just for reading data, multiprocessing would be pretty bad.

Reply:
I'm not 100% sure on this stuff, this is just what I gathered from reading the multiprocessing documentation and experimentation. We probably want to use an actual Lock() primitive, just not sure where it would get instantiated/used and I don't want to deal with that rn.

Reply:
Ah I see. So if you set something to both variables, in each function, then the problem wouldn't occur (not that I want you to, just trying to understand).

Okay so one variable works. Can you still not use the job type and still use another variable (maybe a new variable called `first_job_finished`)?

Also just going to approve now since everything else looks good.